flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/BufferResponseDecoder.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.netty;
+
+import org.apache.flink.shaded.netty4.io.netty.buffer.ByteBuf;
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelHandlerContext;
+
+import static org.apache.flink.runtime.io.network.netty.NettyMessage.BufferResponse;
+import static org.apache.flink.runtime.io.network.netty.NettyMessage.BufferResponse.MESSAGE_HEADER_LENGTH;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * The decoder for {@link BufferResponse}.
+ */
+class BufferResponseDecoder extends NettyMessageDecoder {
+
+	/** The Buffer allocator. */
+	private final NetworkBufferAllocator allocator;
+
+	/** The accumulation buffer of message header. */
+	private ByteBuf messageHeaderBuffer;
+
+	/**
+	 * The BufferResponse message that has its message header decoded, but still
```

Comment:
This description is only for one of the states in processing. Sometimes the header has not been fulfilled yet.